### PR TITLE
android: Migrate from `ndk-glue` to `ndk-context`

### PR DESCRIPTION
**Pull request by MarijnS95**
`ndk-glue` suffers one fatal flaw: it's "only" supposed to be used by the crate providing `fn main()` and only supposed to end up in the dependency graph once as it has `static` globals which get duplicated across versions.

In the current case with `winit 0.26` still on `ndk-glue 0.5` but `cpal` on `ndk-glue 0.6` it'll always panic in `fn native_activity()` as the `static` globals on this version is not initialized.

Introducing `ndk-context`: a crate that holds these `static`s, with the intention/premise to not see a breaking release /ever/ and make this a problem of the past.  The crate is currently initialized with the VM and Android Context on `ndk-glue` 0.5.1 and 0.6.1 (0.4.1 pending) making it compatible with whatever is current, and the possibility for backporting to older `ndk-glue` versions too.

See also:
https://github.com/rust-windowing/android-ndk-rs/issues/211
https://github.com/rust-windowing/android-ndk-rs/pull/223

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -19,6 +19,7 @@ anyhow = "1.0.12"
 hound = "3.4"
 ringbuf = "0.2"
 clap = { version = "3", default-features = false, features = ["std"] }
+ndk-glue = "0.6"
```
Comment (est31) [inline Cargo.toml:24]: Why is this needed?

Comment (MarijnS95) [reply Cargo.toml:24]: Moving it from normal dependencies to `dev-dependencies` for `examples/android.rs`.

Comment (est31) [reply Cargo.toml:24]: Normal dependencies can be accessed from the examples as well.

Comment (MarijnS95) [reply Cargo.toml:24]: The main point is removing `ndk-glue` from the normal dependencies and replacing it with `ndk-context`.

**Review by est31 — APPROVED** [review]
Thanks!

**Comment by est31** [comment]
Hmm wait it failed in CI:
```
error: custom attribute panicked
 --> examples/android.rs:8:35
  |
8 | #[cfg_attr(target_os = "android", ndk_glue::main(backtrace = "full"))]
  |                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: `"ndk-glue"` is not a valid identifier

warning: unused import: `HostTrait`
 --> examples/android.rs:6:33
  |
6 | use cpal::traits::{DeviceTrait, HostTrait, StreamTrait};
  |                                 ^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: `cpal` (example "android") generated 1 warning
error: could not compile `cpal` due to previous error; 1 warning emitted
```

**Comment by MarijnS95** [comment]
It seems that `target.'cfg(target_os = "android")'` doesn't work for `dev-dependencies`? Removing it makes the build succeed again (but breaks other targets for obvious reasons). https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#development-dependencies contains such an example with `cfg()` for `dev-dependencies`, any idea what I'm missing?

**Comment by MarijnS95** [comment]
This looks to be an odd issues inside our macros, investigating...

**Comment by est31** [comment]
This might be because cpal is still on the 2015 edition? The macro might be emitting a wrong span that points to cpal's sources so rustc treats it as edition 2015 code.

**Comment by MarijnS95** [comment]
https://github.com/bkchr/proc-macro-crate/pull/15

I'll think about removing the erroneous fallback in `ndk-macro` too.

**Comment by MarijnS95** [comment]
@est31 the fix to `proc-macro-crate` has been merged and released, re-running the CI should now succeed!

(The workaround has also been removed from `ndk-macro`, but since it's not critical I haven't bothered to create a release yet)

**Comment by MarijnS95** [comment]
Well, I'm not sure how this happened: I made sure to test the fix locally, yet it seems we're now hit by https://github.com/bkchr/proc-macro-crate/pull/16... Not sure how I got it working without that at some point.

**Comment by est31** [comment]
@MarijnS95 it's alright, take your time. I'm impressed by how quickly you figured out the cause of the problem and came up with a solution!

**Comment by MarijnS95** [comment]
@est31 I just don't like making PRs (to `proc-macro-crate`) that are "half finished" and didn't actually fix the issue, shows that my testing methodology was/is messed up somewhere :flushed: 

Anyway, the second fix is merged and released now too. I double checked it locally and the CI should not fail on it now :crossed_fingers:

**Comment by MarijnS95** [comment]
Looks like we're all set now, the current issues also appear on `master`. Perhaps these CI steps should be made infallible so that it looks a bit nicer in the git log, instead of seeing everything failed?

**Comment by est31** [comment]
Thanks again, please write if you want a release. As for the CI failures, yeah maybe the three failling jobs should be commented out or be allowed to fail. Ideally at least the windows job would be fixed tho. PRs welcome :).

**Comment by MarijnS95** [comment]
No need for a release from my side yet, just propagating our `android-ndk-rs` changes through the most popular crates in the ecosystem. Perhaps @zarik5?

**Comment by zarik5** [comment]
I don't have any hurry, I'm keeping my cpal and oboe forks for my project. But once both this PR and the oboe one hit a new release I will be able to remove some duplicate code, thanks to `ndk_context::initialize_android_context()` (my project can't use ndk-glue yet)
